include/swift/SIL/SILBasicBlock.h

```diff
@@ -488,4 +488,126 @@ struct ilist_traits<::swift::SILBasicBlock>
+///
+/// Warning: This is invalid for CondBranchInst arguments. Clients assume that
+/// any instructions inserted at the phi argument is post-dominated by that phi
+/// argument. This warning can be removed once the SILVerifier fully prohibits
+/// CondBranchInst arguments at all SIL stages.
+struct PhiOperand {
+  SILBasicBlock *predBlock = nullptr;
+  unsigned argIndex = 0;
+
+  PhiOperand() = default;
+
+  PhiOperand(Operand *operand) {
+    auto *branch = dyn_cast<BranchInst>(operand->getUser());
+    if (!branch)
+      return;
+
+    predBlock = branch->getParent();
+    argIndex = operand->getOperandNumber();
+  }
+
+  bool operator==(PhiOperand other) const {
+    return predBlock == other.predBlock && argIndex == other.argIndex;
+  }
+
+  Operand *getOperand() const {
+    auto *branch = cast<BranchInst>(predBlock->getTerminator());
+    return &branch->getAllOperands()[argIndex];
+  }
+
+  SILPhiArgument *getValue() const {
+    auto *branch = cast<BranchInst>(predBlock->getTerminator());
+    return cast<SILPhiArgument>(branch->getDestBB()->getArgument(argIndex));
+  }
+
+  operator Operand *() const { return getOperand(); }
+  Operand *operator*() const { return getOperand(); }
+  Operand *operator->() const { return getOperand(); }
+};
+
+/// Represent a phi value without referencing the SILValue, which is invalidated
+/// by adding new, unrelated phi values. Because this only stores a block
+/// pointer, it remains valid as long as the CFG is immutable and the index of
+/// the phi value does not change.
+struct PhiValue {
+  SILBasicBlock *phiBlock = nullptr;
+  unsigned argIndex = 0;
+
+  PhiValue() = default;
+
+  PhiValue(SILValue value) {
+    auto *blockArg = dyn_cast<SILPhiArgument>(value);
+    if (!blockArg || !blockArg->isPhiArgument())
+      return;
+
+    phiBlock = blockArg->getParent();
+    argIndex = blockArg->getIndex();
+  }
+
+  bool operator==(PhiValue other) const {
+    return phiBlock == other.phiBlock && argIndex == other.argIndex;
+  }
+
+  SILPhiArgument *getValue() const {
+    return cast<SILPhiArgument>(phiBlock->getArgument(argIndex));
+  }
+
+  operator SILValue() const { return getValue(); }
+  SILValue operator*() const { return getValue(); }
+  SILValue operator->() const { return getValue(); }
+};
+
+} // namespace swift
+
+namespace llvm {
+
+template <> struct DenseMapInfo<swift::PhiOperand> {
+  static swift::PhiOperand getEmptyKey() { return swift::PhiOperand(); }
+  static swift::PhiOperand getTombstoneKey() {
+    swift::PhiOperand phiOper;
+    phiOper.predBlock =
+        llvm::DenseMapInfo<swift::SILBasicBlock *>::getTombstoneKey();
+    return phiOper;
+  }
+  static unsigned getHashValue(swift::PhiOperand phiOper) {
+    return llvm::hash_combine(phiOper.predBlock, phiOper.argIndex);
+  }
+  static bool isEqual(swift::PhiOperand lhs, swift::PhiOperand rhs) {
+    return lhs == rhs;
+  }
+};
+
+template <> struct DenseMapInfo<swift::PhiValue> {
```

Comment:
Can you rebase this so you get my change to .clang-format +use git-clang-format so template is on the next line. = ).